spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/keras/KerasLayer.scala

```diff
@@ -0,0 +1,435 @@
+
+  override def gradInput_=(value: A): Unit = {
+    labor._gradInput = value
+  }
+
+  override def inputShapeValue_=(value: Shape): Unit = {
+    labor._inputShapeValue = value
+  }
+
+  override def outputShapeValue_=(value: ArrayBuffer[Shape]): Unit = {
+   labor._outputShapeValue = value
+  }
+  // scalastyle:on
+
+  override def compatibleWithKeras(): Boolean = true
+
+  override def compatibleWithTorch(): Boolean = false
+
+  override def getInputShape(): Shape = {
+    if (batchInputShape != null) {
+      batchInputShape
+    } else if (this.labor == null) {
+      null
+    } else {
+      this.labor.getInputShape()
+    }
+  }
+
+  override def computeOutputShape(inputShape: Shape): Shape = {
+    this.labor.computeOutputShape(inputShape)
+  }
+
+  override def getOutputShape(): Shape = labor.getOutputShape()
+
+  override def build(inputShape: Shape): Shape = {
+    labor = doBuild(inputShape)
+    if (!labor.isBuilt) {
+      labor.build(inputShape)
+    }
+    labor.isBuilt = true
+    isBuilt = true
+    labor.getOutputShape()
+  }
+
+  def doBuild(inputShape: Shape): AbstractModule[A, B, T]
+
+  /**
+   * Get the scale of gradientWeight
+   */
+  override def getScaleW(): Double = labor.getScaleW()
+
+  /**
+   * Get the scale of gradientBias
+   */
+  override def getScaleB(): Double = labor.getScaleB()
+
+  /**
+   * Set the scale of gradientWeight
+   *
+   * @param w the value of the scale of gradientWeight
+   * @return this
+   */
+  override def setScaleW(w: Double): this.type = {
+    labor.setScaleW(w)
+    this
+  }
+
+  /**
+   * Set the scale of gradientBias
+   *
+   * @param b the value of the scale of gradientBias
+   * @return this
+   */
+  override def setScaleB(b: Double): this.type = {
+    labor.setScaleB(b)
+    this
+  }
+
+  /**
+   * Clear cached activities to save storage space or network bandwidth. Note that we use
+   * Tensor.set to keep some information like tensor share
+   *
+   * The subclass should override this method if it allocate some extra resource, and call the
+   * super.clearState in the override method
+   *
+   * @return
+   */
+  override def clearState() : this.type = {
+    labor.clearState()
+    this
+  }
+
+  override def toString(): String = {
+    val prefix = getPrintName()
+    val details = if (labor != null) {"<" + labor.toString + ">"} else ""
+    prefix + details
+  }
+
+
+  override def getTimes(): Array[(AbstractModule[_ <: Activity, _ <: Activity, T], Long, Long)] = {
+    labor.getTimes()
+  }
+
+  override def resetTimes(): Unit = labor.resetTimes()
+
+  /**
+   * freeze the module,
+   * i.e. their parameters(weight/bias, if exists) are not changed in training process
+   * if names is not empty,
+   * set an array of layers that match the given ```names``` to be "freezed",
+   *
+   * @param names an array of layer names
+   * @return current graph model
+   */
+  override def freeze(names: String*): this.type = {
```

Comment:
I think you needn't overide such method. As long as make KerasLayer as a subclasss of Container. Maybe it's better to manage all such overide in Container?

Reply:
this is to ensure that `labor` can also be `freeze` if this method is invoked. can it be auto completed by subclass a contanier?

Reply:
After a second thought, KerasLayer indeed should be a subclass of Container instead of AbstractModule to save some boilerplate code.